Objects/dictobject.c

```diff
@@ -5246,7 +5278,11 @@ PyTypeObject PyDictKeys_Type = {
 static PyObject *
 dictkeys_new(PyObject *dict, PyObject *Py_UNUSED(ignored))
 {
-    return _PyDictView_New(dict, &PyDictKeys_Type);
+    PyObject *view;
+    Py_BEGIN_CRITICAL_SECTION(dict);
+    view = _PyDictView_New(dict, &PyDictKeys_Type);
```

Comment:
No critical section here. Creating the view objects, like `_PyDictView_New` (and the reverse iterator above) do not access any of the dictionary internals so they don't need to lock the dictionary.